axelor-base/src/main/java/com/axelor/apps/base/service/tax/TaxService.java

```diff
@@ -192,4 +199,102 @@ public Set<Tax> getTaxSet(Set<TaxLine> taxLineSet) {
     }
     return taxLineSet.stream().map(TaxLine::getTax).collect(Collectors.toSet());
   }
+
+  /**
+   * TaxLines contain taxes, this method reuses checkTaxesNotOnlyNonDeductibleTaxes(Set<Tax> taxes)
+   *
+   * @param taxLines
+   * @return
+   */
+  public void checkTaxLinesNotOnlyNonDeductibleTaxes(Set<TaxLine> taxLines) throws AxelorException {
+    if (taxLines == null || taxLines.isEmpty()) {
+      return;
+    }
+    Set<Tax> taxes = new HashSet<>();
+    for (TaxLine taxLine : taxLines) {
+      taxes.add(taxLine.getTax());
+    }
+    if (!checkTaxesNotOnlyNonDeductibleTaxes(taxes)) {
+      throw new AxelorException(
+          TraceBackRepository.CATEGORY_INCONSISTENCY,
+          I18n.get(BaseExceptionMessage.TAX_ONLY_NON_DEDUCTIBLE_TAXES_SELECTED_ERROR));
+    }
+  }
+
+  public boolean checkTaxesNotOnlyNonDeductibleTaxes(Set<Tax> taxes) {
+    if (taxes == null || taxes.isEmpty()) {
+      return true;
+    }
+    int countDeductibleTaxes = 0;
+    int countNonDeductibleTaxes = 0;
+    for (Tax tax : taxes) {
+      Boolean isNonDeductibleTax = tax.getIsNonDeductibleTax();
+      if (isNonDeductibleTax) {
+        countNonDeductibleTaxes++;
+      } else {
+        countDeductibleTaxes++;
+      }
+    }
+    if (countDeductibleTaxes == 0 && countNonDeductibleTaxes > 0) {
+      return false;
+    }
+    return true;
```

Comment:
This would be easier to do it with something like 
!taxes.stream().allMatch(Tax::getIsNonDeductibleTax)